src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -133,85 +134,100 @@ gl_sframe augment_data(const gl_sframe &data,
   size_t total_augmented_rows = data.size() * backgrounds_size;
   table_printer table( { {"Images Augmented", 0}, {"Elapsed Time", 0}, {"Percent Complete", 0} } );
   if (verbose) {
-    logprogress_stream << "Augmenting input images using " << backgrounds.size() << " background images." << std::endl;
+    logprogress_stream << "Augmenting input images using " << backgrounds_size << " background images." << std::endl;
     table.print_header();
   }
-
+  std::vector<std::string> output_column_names = {"image", "annotation"};
+  std::vector<flex_type_enum> output_column_types = {flex_type_enum::IMAGE, flex_type_enum::DICT};
+  gl_sframe_writer output_writer(output_column_names, output_column_types);  
   auto column_index_map = generate_column_index_map(data.column_names());
-  std::vector<flexible_type> annotations;
-  std::vector<flexible_type> images;
-  int input_row_index = -1;
-  for (const auto& row: data.range_iterator()) {
-    input_row_index++;
-    flex_image object = row[column_index_map[image_column_name]].to<flex_image>();
-    std::string label = row[column_index_map[target_column_name]].to<flex_string>();
-    if (!(object.is_decoded())) {
-      decode_image_inplace(object);
-    }
-    int row_number = -1;
-    for (const auto& background_ft: backgrounds.range_iterator()) {
+  size_t image_column_index = column_index_map[image_column_name];
+  size_t target_column_index = column_index_map[target_column_name];
+  DASSERT_EQ(data[image_column_name].dtype(), flex_type_enum::IMAGE);
+  DASSERT_EQ(data[target_column_name].dtype(), flex_type_enum::STRING);
+  size_t nsegments = output_writer.num_segments();
+  std::atomic<size_t> augmented_counter(0);
+
+  /* Split all backgrounds into as many chunks as there are cores available
+   * (= nsegments), and create augmented images in parallel.
+   */
+  parallel_for(0, nsegments, [&](size_t segment_id){
+    size_t segment_start = (segment_id * backgrounds.size()) / nsegments;
+    size_t segment_end   = ((segment_id+1) * backgrounds.size()) / nsegments;
+    size_t row_number=segment_start;
+    for (const auto& background_ft: backgrounds.range_iterator(segment_start, segment_end)) {
       row_number++;
-      flex_image flex_background = background_ft.to<flex_image>();
-      if (!(flex_background.is_decoded())) {
-        decode_image_inplace(flex_background);
+      for (const auto& row: data.range_iterator()) {
+        // go through all the starter images and create augmented images for
+        // all starter images and the respective chunk of background images
+        flex_image object = row[image_column_index].to<flex_image>();
+        std::string label = row[target_column_index].to<flex_string>();
+        if (!(object.is_decoded())) {
```

Comment:
Any reason not to decode the start images upfront, before the `parallel_for`? Instead of requiring each thread to create its own local decoded version?

Reply:
Fixed.